src/devices/video/dp8510.cpp

```diff
@@ -0,0 +1,211 @@
+		(data & CONTROL_FS) >> 12,
+		(data & CONTROL_SN) >> 8,
+		(data & CONTROL_LM) >> 4,
+		(data & CONTROL_RM) >> 0,
+		machine().describe_context());
+
+	m_control = data;
+
+	// expand function select bits to 16 bit size for convenience
+	m_f0 = (data & FS_0) ? ~u16(0) : 0;
+	m_f1 = (data & FS_1) ? ~u16(0) : 0;
+	m_f2 = (data & FS_2) ? ~u16(0) : 0;
+	m_f3 = (data & FS_3) ? ~u16(0) : 0;
+}
+
+void dp8510_device::source_w(const u16 data, const bool fifo_write)
+{
+	if (fifo_write)
+	{
+		// execute barrel shift
+		const u8 shift_amount = (m_control & CONTROL_SN) >> 8;
+		const u16 shift_result = m_barrel_input_select
+			? (data << shift_amount) | (m_barrel_input_latch >> (16 - shift_amount))
+			: (m_barrel_input_latch << shift_amount) | (data >> (16 - shift_amount));
+
+		LOG("source_w 0x%04x shifted data 0x%04x written to fifo\n", data, shift_result);
+
+		// write result to fifo
+		fifo_w(shift_result);
+	}
+	else
+		LOG("source_w barrel input latch 0x%04x\n", data);
+
+	// save data in input latch
+	m_barrel_input_latch = data;
+}
+
+u16 dp8510_device::logic_unit(const u16 src, const u16 dst) const
+{
+	// compute logic unit result
+	const u16 result =
+		(m_f3 & src & dst) +
+		(m_f2 & src & ~dst) +
+		(m_f1 & ~src & dst) +
+		(m_f0 & ~src & ~dst);
+
+	LOG("logic_unit operation %s source 0x%04x destination 0x%04x result 0x%04x\n",
+		BITBLT_OP[(m_control & CONTROL_FS) >> 12], src, dst, result);
+
+	return result;
+}
+
+void dp8510_device::destination_w(const u16 data, const bool fifo_write)
+{
+	// fetch the source from the pixel data input or fifo
+	const u16 source = m_line_drawing ? m_pixel_data_input ? 0xffff : 0 : m_fifo_output_latch = fifo_r();
+
+	// compute the result
+	u16 result = logic_unit(source, data);
+
+	// apply left mask
+	if (m_left_mask)
+	{
+		const u16 left_mask = 0xffffU >> ((m_control & CONTROL_LM) >> 4);
+		if (left_mask != 0xffff)
+			LOG("destination_w left mask 0x%04x applied\n", left_mask);
+
+		result = (result & left_mask) | (data & ~left_mask);
+	}
+
+	// apply right mask
+	if (m_right_mask)
+	{
+		const u16 right_mask = s16(0x8000) >> (m_control & CONTROL_RM);
+		if (right_mask != 0xffff)
+			LOG("destination_w right mask 0x%04x applied\n", right_mask);
+
+		result = (result & right_mask) | (data & ~right_mask);
+	}
+
+	// latch output
+	m_logic_output = result;
+
+	// write the result back to the fifo
+	if (fifo_write)
+	{
+		LOG("destination_w masked data 0x%04x written to fifo\n", m_logic_output);
+		fifo_w(m_logic_output);
+	}
+	else
+		LOG("destination_w masked data 0x%04x latched\n", m_logic_output);
+}
+
+u16 dp8510_device::output_r() const
+{
+	// return output from fifo or logic unit
+	if (m_data_output_select)
+		return m_fifo_output_latch;
+	else
+		return m_logic_output;
+}
+
+u8 dp8510_device::pixel_r(const offs_t pixel_address) const
+{
+	return (m_logic_output >> (~pixel_address & 0xf)) & 0x1;
+}
+
+u16 dp8510_device::fifo_r()
+{
```

Comment:
Does this ever get called from the debugger?  Should it check `machine().side_effect_disabled()` and suppress side effects?